### DISCUSSION TITLE
Upsert and attributes that are composites of the table’s primary partition or sort keys

### DISCUSSION THREAD

**Discussion by dls314** [tywalch/electrodb#442] [Q&A] [answered]
In the upsert documentation, there is this note

> Note: upsert will apply a dynamodb set operation on all attributes provided. If an attribute is configured as readOnly, ElectroDB will apply an if_not_exists condition to prevent overwriting the existing value. The exception case is if the attribute is a composite of an index, in which case it will NOT be wrapped with if_not_exists()

I wanted to ask about the exception case if the attribute in question is a composite of the table’s primary partition or sort keys.

Since those keys can't be updated, I would expect that the exception doesn't apply, and ElectroDB would apply an if_not_exists condition for those attributes that are composites of the table’s primary partition or sort keys if they are marked as readOnly.

While trying to figure that out, I don't think that https://github.com/tywalch/electrodb/blob/master/src/entity.js#L2466-L2506 has an exception for attributes that are composites of the table’s primary partition or sort keys if they are marked as readOnly.

Is there a good example of how I could add if_not_exists conditions for attributes that are composites of the table’s primary partition or sort keys during an upsert?

**Comment by tywalch** [comment, marked as answer]
I believe the reason for this is because the primary PK and SK are also the item's identifiers. So if the item was found with the composite values provided (i.e. falling into the "update" scenario of the "upsert") the composite value on the stored can be expected to match the one provided. 

ElectroDB has strict constraints to keep keys and their composites in sync (failure to do so would be a serious bug), so the attribute values should always match the indexes value. As such, this operation (without if_not_exists()) is safe at best and auto healing at worst; ideally not worst though 😅

Hope this helps!

**Comment by dls314** [comment]
I have a similar question about ensuring read only semantics during an upsert for attributes that are composites of partition or sort keys for secondary indexes.

From what I can see, it's maybe not possible to do so right now using upsert, but I'm unsure if this is true or not.

**Reply by tywalch** [reply]
> From what I can see, it's maybe not possible to do so right now using upsert, but I'm unsure if this is true or not.

@dls314 can you demonstrate what you're hoping to perform?
